gh-131831: Implement PEP 758 – Allow except and except* expressions without parentheses (#131833)
python/cpython: Doc/whatsnew/3.14.rst

New features
============

> PEP 758 – Allow except and except* expressions without parentheses

Reviewer: ```suggestion
.. _whatsnew314-pep758:

PEP 758 – Allow except and except* expressions without parentheses
```

------------------------------------------------------------------

> The :keyword:`except` and :keyword:`except*` expressions now allow

Author: ```suggestion
The :keyword:`except` expressions now allow
```

> The :keyword:`except` and :keyword:`except*` expressions now allow

Reviewer 2: ```suggestion
The :keyword:`except` and :keyword:`except* <except_star>` expressions now allow
```

> The :keyword:`except` and :keyword:`except*` expressions now allow

Reviewer 3: ```suggestion
The :keyword:`except` and ``except*`` expressions now allow
```

`except*` is not a keyword. This will cause the docs build and workflow to fail. We can using ``except*``
 https://github.com/python/cpython/blob/main/Doc%2Ftutorial%2Ferrors.rst#L557-L557

Reviewer: This was suggested here: https://github.com/python/cpython/pull/131833#discussion_r2019158994

parentheses to be omitted when there is only one exception type.
For example the following expressions are now valid:

>   .. code-block:: python

Reviewer 2: Indenting this directive causes the whole block to be indented, ignore if that's intended
```suggestion
.. code-block:: python
```

> .. code-block:: python
> 
>       try:
>           release_new_sleep_token_album()
>       except AlbumNotFound, SongsTooGoodToBeReleased:
>           print("Sorry, no new album this year.")
> 
>       try:
>           release_new_sleep_token_album()
>       except* AlbumNotFound, SongsTooGoodToBeReleased:
>           print("Sorry, no new album this year.")

Reviewer: ```suggestion
.. code-block:: python

   try:
       release_new_sleep_token_album()
   except AlbumNotFound, SongsTooGoodToBeReleased:
       print("Sorry, no new album this year.")

   try:
       release_new_sleep_token_album()
   except* AlbumNotFound, SongsTooGoodToBeReleased:
       print("Sorry, no new album this year.")
```

Check :pep:`758` for more details.

> (Contributed by Pablo Galindo and Brett Cannin in :issue:`131831`.)

Reviewer 4: ```suggestion
(Contributed by Pablo Galindo and Brett Cannon in :issue:`131831`.)
```

Reviewer 5: The typo doesn't seem to be resolved.

> (Contributed by Pablo Galindo and Brett Cannon in :issue:`131831`.)
> 

Reviewer: ```suggestion
(Contributed by Pablo Galindo and Brett Cannon in :gh:`131831`.)

```

New features
============

.. _whatsnew314-pep758:

PEP 758 – Allow except and except* expressions without parentheses
------------------------------------------------------------------

The :keyword:`except` and :keyword:`except* <except_star>` expressions now allow

> parentheses to be omitted when there is only one exception type.

Reviewer: I missed it in the first place here, but it should be "multiple exception types" here right?

> parentheses to be omitted when there is only one exception type.

Author: ```suggestion
parentheses to be omitted when there are multiple exception types and the ``as`` clause is not used.
```

For example the following expressions are now valid:

.. code-block:: python

   try:
       release_new_sleep_token_album()
   except AlbumNotFound, SongsTooGoodToBeReleased:
       print("Sorry, no new album this year.")

> 

Reviewer 5: You may add a short sentence to highlight that it's an exception group, since I didn't notice `*` in `except*` when I read it first.

Author: Done

> .. code-block:: python
> 
>    try:
>        release_new_sleep_token_album()
>    except AlbumNotFound, SongsTooGoodToBeReleased:
>        print("Sorry, no new album this year.")
> 
>    try:
>        release_new_sleep_token_album()
>    except* AlbumNotFound, SongsTooGoodToBeReleased:
>        print("Sorry, no new album this year.")

Author: ```suggestion
.. code-block:: python

   try:
       release_new_sleep_token_album()
   except AlbumNotFound, SongsTooGoodToBeReleased:
       print("Sorry, no new album this year.")

    # The same applies to except* (for exception groups):
   try:
       release_new_sleep_token_album()
   except* AlbumNotFound, SongsTooGoodToBeReleased:
       print("Sorry, no new album this year.")
```
